### PR TITLE
Additional DST related code changes

### PR DESCRIPTION
**Pull request by gordonb3**
Addresses time calculations in NotificationHelper and Scheduler

### CODE REVIEW

File: main/Scheduler.cpp [hunk 1]
```diff
@@ -367,17 +367,29 @@ void CScheduler::SetSunRiseSetTimers(const std::string &sSunRise, const std::str
 		min = atoi(sSunRise.substr(3, 2).c_str());
 		sec = atoi(sSunRise.substr(6, 2).c_str());
 
-		ltime.tm_hour = hour;
-		ltime.tm_min = min;
-		ltime.tm_sec = sec;
-		temptime = mktime(&ltime);
+		int isdst = ltime.tm_isdst;
+		bool goodtime = false;
+		while (!goodtime) {
```
Comment (gizmocuz) [inline main/Scheduler.cpp:372]: Maybe we should make this a function as it's used in quite a lot of places

**Comment by gordonb3** [comment]
True, that's why I created the function for converting the SQL datetime string to time_t. It makes the code a lot easier to read. The content of the other loops is not always the exact same thing though and I felt a bit reluctant to change localtime_r.h again. It shouldn't be too hard to build a (set of) function(s) around this though. Say:
- void makeMidnightTime(time_t &time, struct tm &result, int year, int month, int day);
- void makeVerifiedTime(time_t &time, struct tm &result, int year, int month, int day, int hour, int minute, int second);
- void makeVerifiedTime(time_t &time, struct tm &result, int year, int month, int day, int hour, int minute, int second, int isdst);

Or:
- time_t makeMidnightTime(struct tm &result, int year, int month, int day);

But I think that last one puts too much focus on the time_t value while in most cases we are mainly interested in the tm struct.

**Comment by gizmocuz** [comment]
Okey...

**Comment by gizmocuz** [comment]
Thanks!

**Comment by EddyK69** [comment]
Please revert the DST changes!

Makes a mess off schedules...

2016-11-17 20:44:12.938 Schedule item started! Name: Selector, Type: Yearly, DevID: 29, Time: 2016-11-17 20:44:12
2016-11-17 20:44:12.939 (Virtual) Light/Switch (Selector)
2016-11-17 20:44:12.948 Schedule item started! Name: Setpoint, Type: Yearly, ThermostatID: 47, Time: 2016-11-17 20:44:12
2016-11-17 20:44:12.949 (Virtual) Thermostat (Setpoint)
2016-11-17 20:44:13.952 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-17 20:44:13
2016-11-17 20:44:13.953 (Virtual) Lighting 2 (Switch 1)
2016-11-17 20:44:13.962 Schedule item started! Name: Selector, Type: Yearly, DevID: 29, Time: 2016-11-17 20:44:13
2016-11-17 20:44:13.963 (Virtual) Light/Switch (Selector)
2016-11-17 20:44:13.972 Schedule item started! Name: Setpoint, Type: Yearly, ThermostatID: 47, Time: 2016-11-17 20:44:13
2016-11-17 20:44:13.973 (Virtual) Thermostat (Setpoint)
2016-11-17 20:44:14.976 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-17 20:44:14
2016-11-17 20:44:14.977 (Virtual) Lighting 2 (Switch 1)
2016-11-17 20:44:14.986 Schedule item started! Name: Selector, Type: Yearly, DevID: 29, Time: 2016-11-17 20:44:14
2016-11-17 20:44:14.987 (Virtual) Light/Switch (Selector)
2016-11-17 20:44:14.996 Schedule item started! Name: Setpoint, Type: Yearly, ThermostatID: 47, Time: 2016-11-17 20:44:14
2016-11-17 20:44:14.997 (Virtual) Thermostat (Setpoint)
2016-11-17 20:44:16.000 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-17 20:44:16
2016-11-17 20:44:16.001 (Virtual) Lighting 2 (Switch 1)
2016-11-17 20:44:16.010 Schedule item started! Name: Selector, Type: Yearly, DevID: 29, Time: 2016-11-17 20:44:16
2016-11-17 20:44:16.011 (Virtual) Light/Switch (Selector)
2016-11-17 20:44:16.021 Schedule item started! Name: Setpoint, Type: Yearly, ThermostatID: 47, Time: 2016-11-17 20:44:16
2016-11-17 20:44:16.021 (Virtual) Thermostat (Setpoint)
2016-11-17 20:44:17.025 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-17 20:44:17
2016-11-17 20:44:17.026 (Virtual) Lighting 2 (Switch 1)
2016-11-17 20:44:17.035 Schedule item started! Name: Selector, Type: Yearly, DevID: 29, Time: 2016-11-17 20:44:17
2016-11-17 20:44:17.036 (Virtual) Light/Switch (Selector)
2016-11-17 20:44:17.045 Schedule item started! Name: Setpoint, Type: Yearly, ThermostatID: 47, Time: 2016-11-17 20:44:17
2016-11-17 20:44:17.046 (Virtual) Thermostat (Setpoint)
2016-11-17 20:44:18.049 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-17 20:44:18
2016-11-17 20:44:18.050 (Virtual) Lighting 2 (Switch 1)
2016-11-17 20:44:18.059 Schedule item started! Name: Selector, Type: Yearly, DevID: 29, Time: 2016-11-17 20:44:18
2016-11-17 20:44:18.060 (Virtual) Light/Switch (Selector)
2016-11-17 20:44:18.069 Schedule item started! Name: Setpoint, Type: Yearly, ThermostatID: 47, Time: 2016-11-17 20:44:18
2016-11-17 20:44:18.070 (Virtual) Thermostat (Setpoint)
2016-11-17 20:44:19.073 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-17 20:44:19
2016-11-17 20:44:19.074 (Virtual) Lighting 2 (Switch 1)
2016-11-17 20:44:19.083 Schedule item started! Name: Selector, Type: Yearly, DevID: 29, Time: 2016-11-17 20:44:19
2016-11-17 20:44:19.084 (Virtual) Light/Switch (Selector)
2016-11-17 20:44:19.093 Schedule item started! Name: Setpoint, Type: Yearly, ThermostatID: 47, Time: 2016-11-17 20:44:19
2016-11-17 20:44:19.094 (Virtual) Thermostat (Setpoint)
2016-11-17 20:44:20.097 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-17 20:44:20
2016-11-17 20:44:20.098 (Virtual) Lighting 2 (Switch 1)
2016-11-17 20:44:20.107 Schedule item started! Name: Selector, Type: Yearly, DevID: 29, Time: 2016-11-17 20:44:20
2016-11-17 20:44:20.108 (Virtual) Light/Switch (Selector)
2016-11-17 20:44:20.117 Schedule item started! Name: Setpoint, Type: Yearly, ThermostatID: 47, Time: 2016-11-17 20:44:20
2016-11-17 20:44:20.118 (Virtual) Thermostat (Setpoint)
2016-11-17 20:44:21.121 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-17 20:44:21
2016-11-17 20:44:21.123 (Virtual) Lighting 2 (Switch 1)
2016-11-17 20:44:21.132 Schedule item started! Name: Selector, Type: Yearly, DevID: 29, Time: 2016-11-17 20:44:21
2016-11-17 20:44:21.134 (Virtual) Light/Switch (Selector)
2016-11-17 20:44:21.145 Schedule item started! Name: Setpoint, Type: Yearly, ThermostatID: 47, Time: 2016-11-17 20:44:21
2016-11-17 20:44:21.145 (Virtual) Thermostat (Setpoint)
2016-11-17 20:44:22.148 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-17 20:44:22
2016-11-17 20:44:22.149 (Virtual) Lighting 2 (Switch 1)
2016-11-17 20:44:22.156 Schedule item started! Name: Selector, Type: Yearly, DevID: 29, Time: 2016-11-17 20:44:22
2016-11-17 20:44:22.156 (Virtual) Light/Switch (Selector)
2016-11-17 20:44:22.231 Schedule item started! Name: Setpoint, Type: Yearly, ThermostatID: 47, Time: 2016-11-17 20:44:22

**Comment by gordonb3** [comment]
Ehhh.... Type Yearly? That does not make sense. I'm also not seeing any of this myself:

`2016-11-17 21:28:09.382 Incoming connection from: 127.0.0.1`
`2016-11-17 21:29:01.598 Schedule item started! Name: Buitenstroom, Type: On Time, DevID: 16, Time: 2016-11-17 21:29:01`
`2016-11-17 22:02:01.270 Schedule item started! Name: Buitenstroom, Type: Odd Day Numbers, DevID: 16, Time: 2016-11-17 22:02:01`

Should note that I'm not running the mainline beta but my own [thread](https://github.com/gordonb3/domoticz/tree/fixDST) which includes all the DST changes but is 61 commits (# b97777b) behind on the official master.

**Comment by EddyK69** [comment]
It occurs on Yearly-type when it's time to execute (example Friday Nov-11 @8:04) is runs on 8:04 but keeps executing every second. I think is it not rescheduling correctly.

[EDIT] Same for type: Yearly (weekday)!

2016-11-18 08:04:01.125 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-18 08:04:01
2016-11-18 08:04:01.126 (Virtual) Lighting 2 (Switch 1)
2016-11-18 08:04:02.134 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-18 08:04:02
2016-11-18 08:04:02.135 (Virtual) Lighting 2 (Switch 1)
2016-11-18 08:04:03.142 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-18 08:04:03
2016-11-18 08:04:03.143 (Virtual) Lighting 2 (Switch 1)
2016-11-18 08:04:04.155 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-18 08:04:04
2016-11-18 08:04:04.156 (Virtual) Lighting 2 (Switch 1)
2016-11-18 08:04:05.166 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-18 08:04:05
2016-11-18 08:04:05.167 (Virtual) Lighting 2 (Switch 1)
2016-11-18 08:04:06.174 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-18 08:04:06
2016-11-18 08:04:06.175 (Virtual) Lighting 2 (Switch 1)
2016-11-18 08:04:07.182 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-18 08:04:07
2016-11-18 08:04:07.183 (Virtual) Lighting 2 (Switch 1)
2016-11-18 08:04:08.190 Schedule item started! Name: Switch 1, Type: Yearly, DevID: 7, Time: 2016-11-18 08:04:08
2016-11-18 08:04:08.191 (Virtual) Lighting 2 (Switch 1)

**Comment by gordonb3** [comment]
I see.

Just verified and there is in fact a missing line that causes the code to ignore the part where the schedule is supposed to schedule for next year.

Pull request #994 

Rob, I think I'll take your suggestion to make that highlighted bit of code a function.

**Comment by gizmocuz** [comment]
Hehe ;)

**Pull request merged**